pkg/operator/k8sutil/deployment.go

```diff
@@ -254,11 +254,17 @@ func addLabel(key, value string, labels map[string]string) {
 	labels[key] = value
 }
 
-func CreateDeployment(name, namespace string, clientset kubernetes.Interface, dep *apps.Deployment) error {
+func CreateDeployment(clientset kubernetes.Interface, name, namespace string, dep *apps.Deployment) error {
 	_, err := clientset.AppsV1().Deployments(namespace).Create(dep)
 	if err != nil {
 		if k8serrors.IsAlreadyExists(err) {
-			_, err = clientset.AppsV1().Deployments(namespace).Update(dep)
+			// deleting the deployment if its already exists to avoid issues
```

Comment:
@travisn the pod will stay in a pending state if I don't delete and create the deployment, let me know is there any other way to fix  it without deleting and recreating
```console
[🎩︎]mrajanna@localhost rook $]kuberc get po -owide
NAME                                            READY   STATUS      RESTARTS   AGE     IP               NODE      NOMINATED NODE   READINESS GATES
csi-cephfsplugin-fh5f4                          3/3     Running     0          12m     192.168.121.14   worker0   <none>           <none>
csi-cephfsplugin-provisioner-59b9cb4464-s6gvv   5/5     Running     0          12m     10.44.0.10       worker1   <none>           <none>
csi-cephfsplugin-provisioner-59b9cb4464-w96ld   5/5     Running     0          11m     10.36.0.3        worker0   <none>           <none>
csi-cephfsplugin-provisioner-6f8fd7854f-qpn6h   0/5     Pending     0          5m51s   <none>           <none>    <none>           <none>

 Type     Reason            Age        From               Message
  ----     ------            ----       ----               -------
  Warning  FailedScheduling  <unknown>  default-scheduler  0/3 nodes are available: 1 node(s) had taints that the pod didn't tolerate, 2 node(s) didn't match pod affinity/anti-affinity.
  Warning  FailedScheduling  <unknown>  default-scheduler  0/3 nodes are available: 1 node(s) had taints that the pod didn't tolerate, 2 node(s) didn't match pod affinity/anti-affinity.
```

Reply:
This is when you were testing on a single-node cluster? Or did the CI hit an issue during upgrade? I'd rather not do something special for upgrade on a single-node cluster since we really don't need to support upgrades on these clusters. Perhaps a simple solution is for the operator to query the number of nodes in the cluster. If 1, set the replica of the deployment to 1 instead of 2. This won't be perfect, for example, if there are taints on nodes, but it will keep the scenario of single-node cluster simple and then we won't see pending pods either.

Reply:
This is multinode cluster you can check node name for pods in above list. 
The pods were in pending state even in multi node cluster

Reply:
Interesting, the deployment must be trying to bring up a new pod before deleting the old one, so the anti-affinity can't be satisfied. There may be a different upgrade setting on the deployment that would allow for old pods to be deleted before starting the new ones. But we may need to keep the delete-and-create strategy. In that case we would want to make sure we only update if there is a change in the pod spec. See [this example](https://github.com/rook/rook/blob/25fd0df893e4ee85bced0d276d9e04ef7e56a362/pkg/operator/k8sutil/deployment.go#L70).

Reply:
sure I can make this check and make its only delete-and-create if there is a difference